# MLPC
***Notice: The number of amino acids in a protein sequence cannot exceed 500 AA***


![MLPC](https://github.com/user-attachments/assets/c4f2b9db-90e3-44f9-9ca2-66c95801b31b)

# Installation

Using conda to create env

```
# By conda
conda env create -f MLPC.yaml

# Activate env
source activate mlpc
```

# How to start
## a.Feature engineering
**1.Property:**

Get your fasta file ready (protein), switch your path
```
Rscript property_feature.R -i protein.fa -o feature_table.txt
## -i input protein fasta! format .fa (.fasta)
## -o output property table! format .txt
```

**2.Sequence and Secondary structure:**

For secondary structure predictions we recommend using the SCRATCH-1D to obtain '.ss8' file

```
python feature_engine.py -f protein.fa -s secondary_structure.ss8 -p property.txt -rf feature_table.txt
## -f protein sequence
## -s ss8 format,secondary structure
## -p property table ('\t')
## -rf feature matrix file (output)
```

## b.Trainer

Training set testing set and validation set are generated by random sampling

```
# For example 'glycosidase'
python random_sample.py -i ./feat_glycosidase.txt -or ./random_sample/feat_glycosidase_1500.txt -oe ./testsets/feat_glycosidase_1500.txt -ov ./valid/feat_glycosidase_1500.txt -kr 1500 -ke 1500 -kv 1500
## -i input feature table (txt,'\t')
## -or output training feature table (after sample, txt,'\t')
## -oe output testing feature table (after sample, txt,'\t')
## -ov output validation feature table (after sample, txt,'\t')
## -kr sample number for training
## -ke sample number for testing
## -kv sample number for validation

# Obtain training sets
cat feat_amidase_1500.txt feat_MES_1500.txt feat_glycosidase_1500.txt feat_other_1500.txt feat_endo-peptidase_827.txt > train_1.txt

# Mlsmote
python mlsmote.py -iX train_1.txt -iy laber_class.txt -o random_sample -n 673
## -iX feature table (txt,'\t')
## -iy laber file (txt, '\t')
## -o output path
## -n number of newly generated sample
```

**Training model:**

```
# Obtain training sets after mlsmote
cat feat_amidase_1500.txt feat_MES_1500.txt feat_glycosidase_1500.txt feat_other_1500.txt X_res.txt > train_2.txt

# Run CNN
python run_CNN.py -iX train_2.txt -iy laber.txt
## -iX feature table (txt,'\t')
## -iy laber file (txt, '\t')

# Run BiGRU
python run_BiGRU.py -iX train_2.txt -iy laber.txt
## -iX feature table (txt,'\t')
## -iy laber file (txt, '\t')
```

## c.Tester

If you want to test model performance, testing below:

```
python tester.py -iX test.txt -iy laber.txt -m model.h5
## -iX feature table for test sets (txt,'\t')
## -iy laber file for test set (txt, '\t')
## -m  model path (.h5)
```

## d.Predictor

If you want to predict new proteins, see below:

```
python predictor.py -iX new_feature.txt -r res.txt -m model.h5
## -iX feature table for new sets (txt,'\t')
## -r  result file (txt,'\t')
## -m  model path (.h5)
```

### Supp.
1. Using Grad_CAM.py to calculate BiGRU layer Grad-CAM position weight which is corresponding to secondary structure sequences

```
python Grad_CAM.py -iX feature.txt -s secondary_structure.ss8 -r /.../Grad_CAM_score_folder/ -m model.h5
## -iX feature table for new sets (txt,'\t')
## -s  ss8 file predicted by "SCRATCH"
## -r  result folder (path)
## -m  model path (.h5)
```

2. Using min_length_ss_score.py to calculate different secondary structure average weight in different length, the main purpose is to check the weight of an entire alpha-helix (or other secondary structure type) 

```
python min_length_ss_score.py -f /.../Grad_CAM_score_folder/ -l 5 -r res.csv
## -f  secondary structure sequence grad cam score file path (generated by Grad_CAM.py)
## -l  min length for secondary structure sequence
## -r  result file
```

3. Using motif_finder.py to search for secondary structure motifs that meets the Grad-CAM position weight cut off, users can customize the cut off and length range of motifs

```
python motif_finder.py -i ./Grad_CAM_score_folder/ -o ./motif.csv -c 0.6 -ml 5 -mu 20
## -f  Grad-CAM score folder, the file example in exampleData (folder) protein_Grad_CAM_1.csv or protein_Grad_CAM_2.csv
## -o  result file (.csv)
## -c  Grad-CAM position weight cut off
## -ml Motif length lower
## -mu Motif length upper
```

# Error tip
**a. Tensorflow is not initialized**
If error:
```
Traceback (most recent call last):
  File "Grad_CAM.py", line 145, in <module>
    predict(feats_dat,h5_model,ss8_file,res_path)
  File "Grad_CAM.py", line 76, in predict
    pooled_grads_value, conv_layer_output_value = iterate([X_test[sample_idx:sample_idx+1]])
  File "/home/xiaorunze/miniconda3/envs/mlpc/lib/python3.6/site-packages/tensorflow/python/keras/backend.py", line 2986, in __call__
    run_metadata=self.run_metadata)
  File "/home/xiaorunze/miniconda3/envs/mlpc/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1439, in __call__
    run_metadata_ptr)
  File "/home/xiaorunze/miniconda3/envs/mlpc/lib/python3.6/site-packages/tensorflow/python/framework/errors_impl.py", line 528, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.FailedPreconditionError: Attempting to use uninitialized value FC1/kernel
         [[{{node FC1/kernel/read}} = Identity[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"](FC1/kernel)]]
```
please add codes into 'Grad_CAM.py', see 'Grad_CAM.py' line 45-47:
```
# 对模型中的变量进行初始化,否则会报错
init = tf.global_variables_initializer()
sess = K.get_session()
sess.run(init)
```

# Model download
Model address:
  Link: https://caiyun.139.com/w/i/2nFZ7HJiqBguh
  Password: ucbu
